BoneCentral/Peripherals/Peripherals.SerialAdaptors/src/Serial.cpp

```diff
@@ -1,63 +1,191 @@
-    input_ = std::make_shared<std::ifstream>("/dev/ttyACM0");
-    output_ = std::make_shared<std::ofstream>("/dev/ttyACM0");
-    arduinoInitialized_ = false;
-    std::thread signalwait([](){ initializeArduino(); });
-}
+std::mutex Serial::serialLock_;
 
-void Serial::writeByte(unsigned short byteValue) {
-    if (arduinoInitialized_)
-    {
-        std::lock_guard<std::mutex> guard(serialLock_);
-        *output_ << char(byteValue & 0xFF) << std::flush;
+Serial::Serial(std::string device) {
+#ifdef DEBUG
+    LOG("\nReceived Device Name: " << device);
+    LOG("\nEntering Serial Debug Mode\n");
+    fd = 0;
+#else
+    if((fd = open(device.c_str(), O_RDWR)) < 0) {
+        LOG("Device failed to open... entering dummy mode.\n");
+        fd = 0;
+        return;
     }
+    
+    configure();
+#endif
+    ackno();
 }
 
-void Serial::writeShort(unsigned short shortValue) {
-    if (arduinoInitialized_)
-    {
-        std::lock_guard<std::mutex> guard(serialLock_);
-        *output_ << char(shortValue >> 8) << char(shortValue & 0xFF) << std::flush;
-    }
+Serial::~Serial() {
+    if(fd != 0) close(fd);
 }
 
-json Serial::readJson() {
-    if (arduinoInitialized_)
-    {
-        std::lock_guard<std::mutex> guard(serialLock_);
-        json arduinoData;
-        *input >> arduinoData;
-        return arduinoData;
+void Serial::configure() {
+    struct termios topts;
+    if(tcgetattr(fd, &topts)) {
+        LOG("Failed to get terminal ios options from file descriptor.\n");
+        throw 1;
+    }
+    if(cfsetispeed(&topts, B115200)) {
+        LOG("Failed to set input baud rate.\n");
+        throw 1;
+    }
+    if(cfsetospeed(&topts, B115200)) {
+        LOG("Failed to set output baud rate.\n");
+        throw 1;
+    }
+    topts.c_cflag &= ~(PARENB|CSTOPB|CSIZE|CRTSCTS);
+    topts.c_cflag |= CS8|CREAD|CLOCAL;
+    topts.c_iflag &= ~(IXON | IXOFF | IXANY);
+    topts.c_lflag &= ~(ICANON | ECHO | ECHOE | ISIG);
+    topts.c_oflag &= ~OPOST;
+    topts.c_cc[VMIN] = 1;
+    topts.c_cc[VTIME] = 0;
+    if(tcsetattr(fd, TCSANOW, &topts)) {
+        LOG("Failed to set terminal ios options for file descriptor.\n");
+        throw 1;
+    }
+    if(tcflush(fd, TCIFLUSH)) {
+        LOG("Failed to flush file descriptor.\n");
+        throw 1;
     }
-    return {};
 }
 
-char Serial::readChar() {
-    if (arduinoInitialized_)
-    {
-        std::lock_guard<std::mutex> guard(serialLock_);
-        char arduinoReading;
-        *input >> arduinoReading;
-        return arduinoReading;
-    }
-    return '';
+void Serial::ackno() {
+    std::string response = readString();
+#ifdef DEBUG
+    LOG("Arduino Message: " << response << "\n");
+#endif
+    writeByte('R');
+}
+
+
+
+
+std::string Serial::readString() {
+#ifdef DEBUG
+    return "Dummy String\0";
+#else
+    if(fd==0) return "";
+    std::stringstream ss;
+    char c[1];
+    
+    do{
+        read(fd, c, 1);
+        ss << c[0];
+    }while(c[0] != '\0');
+    return ss.str();
+#endif
```

Comment:
This morning I read up on professional style guides (Google C++ Style Guide) and they are in agreement with you.

https://google.github.io/styleguide/cppguide.html#Preprocessor_Macros

I will try to find alternative approaches that won't litter the code with preprocessor directives but also won't double the code base either.

I will ommit any UNO related code since our project deals solely with the Due.

The problem with implementing two classes (Due and Debug) is that the primary purpose of the debug mode was to actively see and "debug" the flow of information in the main code.

Serial here, of course, probably could have its own "dummy" object, but a preprocessor directive (or an update to cmake) will still be required.